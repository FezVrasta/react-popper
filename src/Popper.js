// @flow
import React, { Component, type Node } from 'react';
import PopperJS, {
  type Placement,
  type Instance as PopperJS$Instance,
  type ReferenceObject,
} from 'popper.js';
import { ManagerContext } from './Manager';
import { unwrapArray } from './utils';

type getRefFn = (?HTMLElement) => void;
type Style = Object;

type ReferenceElement = ReferenceObject | HTMLElement | null;

type RenderProp = ({|
  ref: getRefFn,
  style: Style,
  placement: ?Placement,
  arrowProps: {
    ref: getRefFn,
    style: Style,
  },
|}) => Node;

type PopperProps = {
  modifiers?: {
    [string]: { order: number, enabled: boolean, fn: Object => Object },
  },
  placement?: Placement,
  eventsEnabled?: boolean,
  referenceElement?: ReferenceElement,
  children: RenderProp,
};

type PopperState = {
  popperNode: ?HTMLElement,
  arrowNode: ?HTMLElement,
  popperInstance: ?PopperJS$Instance,
  data: ?Object,
};

const initialStyle = {
  position: 'absolute',
  opacity: 0,
  pointerEvents: 'none',
};

const initialArrowStyle = {};

export class Popper extends Component<PopperProps, PopperState> {
  static defaultProps = {
    placement: 'bottom',
    eventsEnabled: true,
    referenceElement: undefined,
  };

  state = {
    popperNode: undefined,
    arrowNode: undefined,
    popperInstance: undefined,
    data: undefined,
  };

  setPopperNode = (popperNode: ?HTMLElement) => this.setState({ popperNode });
  setArrowNode = (arrowNode: ?HTMLElement) => this.setState({ arrowNode });

  updateStateModifier = {
    enabled: true,
    order: 900,
    fn: (data: Object) => {
      this.setState({ data });
      return data;
    },
  };

  getOptions = () => ({
    placement: this.props.placement,
    eventsEnabled: this.props.eventsEnabled,
    modifiers: {
      ...this.props.modifiers,
      arrow: {
        enabled: !!this.state.arrowNode,
        element: this.state.arrowNode,
      },
      applyStyle: { enabled: false },
      updateStateModifier: this.updateStateModifier,
    },
  });

  getPopperStyle = () =>
    !this.state.popperNode || !this.state.data
      ? initialStyle
      : {
          position: this.state.data.offsets.popper.position,
          ...this.state.data.styles,
        };

  getPopperPlacement = () =>
    !this.state.data ? undefined : this.state.data.placement;

  getArrowStyle = () =>
    !this.state.arrowNode || !this.state.data
      ? initialArrowStyle
      : this.state.data.arrowStyles;

  initPopperInstance = () => {
    const { referenceElement } = this.props;
    const { popperNode, popperInstance } = this.state;
    if (referenceElement && popperNode && !popperInstance) {
      const popperInstance = new PopperJS(
        referenceElement,
        popperNode,
        this.getOptions()
      );
      this.setState({ popperInstance });
      return true;
    }
    return false;
  };

  destroyPopperInstance = (callback: () => boolean) => {
    if (this.state.popperInstance) {
      this.state.popperInstance.destroy();
    }
    this.setState({ popperInstance: undefined }, callback);
  };

  updatePopperInstance = () => {
    if (this.state.popperInstance) {
      this.destroyPopperInstance(() => this.initPopperInstance());
    }
  };

  componentDidUpdate(prevProps: PopperProps, prevState: PopperState) {
    // If needed, initialize the Popper.js instance
    // it will return `true` if it initialized a new instance, or `false` otherwise
    // if it returns `false`, we make sure Popper props haven't changed, and update
    // the Popper.js instance if needed
    if (!this.initPopperInstance()) {
      // If the Popper.js options have changed, update the instance (destroy + create)
      if (
        this.props.placement !== prevProps.placement ||
        this.props.eventsEnabled !== prevProps.eventsEnabled ||
        this.state.arrowNode !== prevState.arrowNode ||
        this.state.popperNode !== prevState.popperNode ||
        this.props.referenceElement !== prevProps.referenceElement
      ) {
        this.updatePopperInstance();
      }
    }
  }

  componentWillUnmount() {
    if (this.state.popperInstance) {
      this.state.popperInstance.destroy();
    }
  }

  render() {
    return unwrapArray(this.props.children)({
      ref: this.setPopperNode,
      style: this.getPopperStyle(),
      placement: this.getPopperPlacement(),
      arrowProps: {
        ref: this.setArrowNode,
        style: this.getArrowStyle(),
      },
    });
  }
}

const placements = PopperJS.placements;
export { placements };

export default (props: PopperProps) => (
  <ManagerContext.Consumer>
    {({ referenceNode }) => (
      <Popper referenceElement={referenceNode} {...props} />
    )}
  </ManagerContext.Consumer>
);
