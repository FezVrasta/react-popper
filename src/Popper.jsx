import React, { Component, createElement } from 'react'
import PropTypes from 'prop-types'
import PopperJS from 'popper.js'
import isEqual from 'is-equal-shallow'

const noop = () => null

class Popper extends Component {
  static contextTypes = {
    popperManager: PropTypes.object.isRequired,
  }

  static childContextTypes = {
    popper: PropTypes.object.isRequired,
  }

  static propTypes = {
    component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    innerRef: PropTypes.func,
    placement: PropTypes.oneOf(PopperJS.placements),
    eventsEnabled: PropTypes.bool,
    modifiers: PropTypes.object,
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  }

  static defaultProps = {
    component: 'div',
    placement: 'bottom',
    eventsEnabled: true,
    modifiers: {},
  }

  state = {}

  subscribers = []

  getChildContext() {
    console.log('getChildContext', this)
    return {
      popper: {
        subscribe: this._subscribe,
        unsubscribe: this._unsubscribe,
        setArrowNode: this._setArrowNode,
      },
    }
  }

  // Add a subcomponent update fn that needs to get updated when we do
  _subscribe = fn => this.subscribers.push(fn)
  // Remove such a fn
  _unsubscribe = fn =>
    (this.subscribers = this.subscribers.filter(s => s !== fn))

  componentDidMount() {
    this._updatePopper()
  }

  componentDidUpdate(lastProps) {
    if (
      lastProps.placement !== this.props.placement ||
      lastProps.eventsEnabled !== this.props.eventsEnabled
    ) {
      this._updatePopper()
    }

    if (this._popper && lastProps.children !== this.props.children) {
      this._popper.scheduleUpdate()
    }
  }

  componentWillUnmount() {
    this._destroyPopper()
  }

  _setArrowNode = node => {
    this._arrowNode = node
  }

  _getTargetNode = () => {
    return this.context.popperManager.getTargetNode()
  }

  _updateFromData = data => {
    if (
      (this.state.data && !isEqual(data.offsets, this.state.data.offsets)) ||
      !this.state.data
    ) {
      this.setState({ data })
      this.subscribers.forEach(s => s(data))
    }
    return data
  }

  _updateStateModifier = {
    enabled: true,
    order: 900,
    fn: this._updateFromData,
  }

  _updatePopper() {
    this._destroyPopper()
    if (this._node) {
      this._createPopper()
    }
  }

  _createPopper() {
    const { placement, eventsEnabled } = this.props
    const modifiers = {
      ...this.props.modifiers,
      applyStyle: { enabled: false },
      updateState: this._updateStateModifier,
    }

    if (this._arrowNode) {
      modifiers.arrow = {
        element: this._arrowNode,
      }
    }

    this._popper = new PopperJS(this._getTargetNode(), this._node, {
      placement,
      eventsEnabled,
      modifiers,
    })

    // schedule an update to make sure everything gets positioned correct
    // after being instantiated
    this._popper.scheduleUpdate()
  }

  _destroyPopper() {
    if (this._popper) {
      this._popper.destroy()
    }
  }

  _getPopperStyle = () => {
    const { data } = this.state

    // If Popper isn't instantiated, hide the popperElement
    // to avoid flash of unstyled content
    if (!this._popper || !data) {
      return {
        position: 'absolute',
        pointerEvents: 'none',
        opacity: 0,
      }
    }

    const { top, left, position } = data.offsets.popper

    return {
      position,
      ...data.styles,
    }
  }

  _getPopperPlacement = () => {
    return !!this.state.data ? this.state.data.placement : undefined
  }

  render() {
    const {
      component,
      innerRef,
      placement,
      eventsEnabled,
      modifiers,
      children,
      ...restProps
    } = this.props

    const popperRef = node => {
      this._node = node
      if (typeof innerRef === 'function') {
        innerRef(node)
      }
    }
    const popperStyle = this._getPopperStyle()
    const popperPlacement = this._getPopperPlacement()

    if (typeof children === 'function') {
      const popperProps = {
        ref: popperRef,
        style: popperStyle,
        ['data-placement']: popperPlacement,
      }
      return children({
        popperProps,
        restProps,
        scheduleUpdate: this._popper && this._popper.scheduleUpdate,
      })
    }

    const componentProps = {
      ...restProps,
      style: {
        ...restProps.style,
        ...popperStyle,
      },
      'data-placement': popperPlacement,
    }

    if (typeof component === 'string') {
      componentProps.ref = popperRef
    } else {
      componentProps.innerRef = popperRef
    }

    return createElement(component, componentProps, children)
  }
}

export default Popper
