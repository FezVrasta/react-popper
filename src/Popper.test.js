// @flow
import React from 'react';
import { mount } from 'enzyme';

// Private API
import { InnerPopper } from './Popper';

const mountPopper = props => new Promise(resolve => {
  const wrapper = mount(
    <InnerPopper onFirstUpdate={() => resolve(wrapper.update())} {...props}>
      {({ref, style, placement, arrowProps}) => (
        <div ref={ref} style={style} data-placement={placement}>
          <div {...arrowProps} />
        </div>
      )}
    </InnerPopper>
  )
});

describe('Popper component', () => {
  it('renders the expected markup', async () => {
    const referenceElement = document.createElement('div');
    const wrapper = await mountPopper({ referenceElement });
    expect(wrapper).toMatchSnapshot();
  });

  it('initializes the Popper.js instance on first update', async () => {
    const referenceElement = document.createElement('div');
    const wrapper = await mountPopper({ referenceElement });
    expect(wrapper.instance().popperInstance).toBeDefined();
  });

  it("doesn't update Popper.js instance on props update if not needed by Popper.js", async () => {
    const referenceElement = document.createElement('div');
    const wrapper = await mountPopper({ referenceElement, placement: 'bottom' });
    const instance = wrapper.instance().popperInstance;

    expect(instance).toBeDefined();

    wrapper.setProps({ placement: 'bottom' });

    expect(wrapper.instance().popperInstance).toBe(instance);
  });

  it('updates Popper.js on explicitly listed props change', async () => {
    const referenceElement = document.createElement('div');
    const wrapper = await mountPopper({ referenceElement });
    const instance = wrapper.instance().popperInstance;
    wrapper.setProps({ placement: 'top' });
    wrapper.update();
    expect(wrapper.instance().popperInstance).toBe(instance);

    await wrapper.instance().popperInstance.update();
    expect(wrapper.instance().popperInstance.state.placement).toBe('top');
  });

  it('does not update Popper.js on generic props change', async () => {
    const referenceElement = document.createElement('div');
    const wrapper = await mountPopper({ referenceElement });
    const instance = wrapper.instance().popperInstance;
    wrapper.setProps({ foo: 'bar' });
    wrapper.update();
    expect(wrapper.instance().popperInstance).toBe(instance);
  });

  it('destroys Popper.js instance on unmount', async () => {
    const referenceElement = document.createElement('div');
    const wrapper = await mountPopper({ referenceElement });
    const component = wrapper.instance();
    wrapper.unmount();
    expect(component.popperInstance).toBeNull();
  });

  it('handles changing refs gracefully', () => {
    const referenceElement = document.createElement('div');
    expect(() =>
      mount(
        <InnerPopper referenceElement={referenceElement}>
          {({ ref, style, placement, arrowProps }) => (
            <div
              ref={current => ref(current)}
              style={style}
              data-placement={placement}
            >
              <div {...arrowProps} ref={current => arrowProps.ref(current)} />
            </div>
          )}
        </InnerPopper>
      )
    ).not.toThrow();
  });

  it('accepts a ref function', () => {
    const myRef = jest.fn();
    const referenceElement = document.createElement('div');
    mount(
      <InnerPopper referenceElement={referenceElement} innerRef={myRef}>
        {({ ref, style, placement}) => (
          <div
            ref={ref}
            style={style}
            data-placement={placement}
          />
        )}
      </InnerPopper>
    )
    expect(myRef).toBeCalled();
  });

  it('accepts a ref object', () => {
    const myRef = React.createRef();
    const referenceElement = document.createElement('div');
    mount(
      <InnerPopper referenceElement={referenceElement} innerRef={myRef}>
        {({ ref, style, placement}) => (
          <div
            ref={ref}
            style={style}
            data-placement={placement}
          />
        )}
      </InnerPopper>
    )
    expect(myRef.current).toBeDefined();
  });

  it('accepts a `referenceElement` property', async () => {
    const virtualReferenceElement = {
      getBoundingClientRect(): any {
        return {
          top: 10,
          left: 10,
          bottom: 20,
          right: 100,
          width: 90,
          height: 10,
        };
      }
    };
    const wrapper = await mountPopper({ referenceElement: virtualReferenceElement });

    expect(wrapper.instance().popperInstance.state.elements.reference).toBe(
      virtualReferenceElement
    );
  });

  it(`should render 3 times when placement is changed`, async () => {
    const referenceElement = document.createElement('div');
    let renderCounter = 0;
    const wrapper = await new Promise(resolve => {
      const wrapper = mount(
        <InnerPopper
          placement="top"
          referenceElement={referenceElement}
          onFirstUpdate={() => resolve(wrapper.update())}
        >
          {({ref, style, placement}) => {
            renderCounter++;
            return <div ref={ref} style={style} data-placement={placement}/>;
          }}
        </InnerPopper>
      )
    });
    expect(renderCounter).toBe(3);

    renderCounter = 0;
    wrapper.setProps({ placement: 'bottom' });
    await wrapper.instance().popperInstance.update();
    expect(renderCounter).toBe(3);
  });
});
