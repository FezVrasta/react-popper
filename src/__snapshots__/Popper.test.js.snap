// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popper component renders the expected markup 1`] = `
<Popper
  eventsEnabled={true}
  placement="bottom"
  referenceElement={<div />}
>
  <div
    style={
      Object {
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
      }
    }
  >
    <div
      style={Object {}}
    />
  </div>
</Popper>
`;
